### PR TITLE
Implement tournaments filtering

### PR DESCRIPTION
**Pull request by alexandrepiveteau**
This PR adds the ability to filter tournaments based on participation and administration criteria. It also adds the user interface to handle finished tournaments, although extra work on the tournament logic is still needed to properly handle this case.

## Deliverables

telegram-cloud-photo-size-4-5789627666494765454-y

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/tournaments/FiltersDialog.kt [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.semantics.contentDescription
+import androidx.compose.ui.semantics.semantics
+import androidx.compose.ui.unit.dp
+import ch.epfl.sdp.mobile.state.LocalLocalizedStrings
+import ch.epfl.sdp.mobile.ui.PawniesColors
+
+/** An interface representing the state of the [FiltersDialog] composable. */
+@Stable
```
Comment (barmettlerl) [inline mobile/src/main/java/ch/epfl/sdp/mobile/ui/tournaments/FiltersDialog.kt:23]: Question: Why is there a `@Stable` Annotation ?

Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/tournaments/FiltersDialog.kt:23]: What was your thought behind it?

**Review by barmettlerl — APPROVED** [review]
Got some questions, but it looks good for me

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt [hunk 2]
```diff
@@ -29,9 +148,12 @@ class TournamentFacade(private val auth: Auth, private val store: Store) {
...
   fun tournaments(user: AuthenticatedUser): Flow<List<Tournament>> {
-    return store.collection(TournamentDocument.Collection).asFlow<TournamentDocument>().map {
-      it.mapNotNull { doc -> doc?.toTournament(user, store) }
-    }
+    val tournaments =
+        store.collection(TournamentDocument.Collection).asFlow<TournamentDocument>().map {
+          it.mapNotNull { doc -> doc?.toTournament(user, store) }
+        }
+    val predicates = tournamentPredicate()
+    return combine(tournaments, predicates) { t, p -> t.filter(p) }
```
[the remark this answers is not in the document]
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt:156]: This applies the predicate `p` to the list of tournaments `t`.

File: mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulHome.kt [hunk 3]
```diff
@@ -68,6 +70,9 @@ private const val TournamentDefaultId = ""
 /** The route associated to new contest button in play screen */
 private const val CreateTournamentRoute = "create_tournament"
 
+/** The route associated with the tournaments filters screen. */
+private const val TournamentFiltersRoute = "tournament_filters"
```
[the remark this answers is not in the document]
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulHome.kt:74]: There isn't one for now, but I agree that there should be one in the future. We might want to tackle this with a bigger refactoring of how top-level navigation is handled, with (maybe) some dedicated `CompositionLocal` which would get passed down to stateful composables.

File: mobile/src/main/java/ch/epfl/sdp/mobile/ui/tournaments/FiltersDialog.kt (hunk 1 above)
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/ui/tournaments/FiltersDialog.kt:23]: The `@Stable` annotation is a contract with the Compose compiler which helps it generate some more optimised code. Long story short, when annotating an interface with `@Stable`, we must make sure all the external properties of the interface implementations will be observable by the composition (aka they use the Snapshot system).

All our viewmodel interfaces have `@Stable` guarantees.

(ref : https://developer.android.com/reference/kotlin/androidx/compose/runtime/Stable)

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt [hunk 4]
```diff
@@ -5,22 +5,141 @@ import ch.epfl.sdp.mobile.application.PoolDocument
...
+  private val keyFilterShowDone: Key<Boolean>
+
+  /** The key which indicates if only participating tournaments should be displayed. */
+  private val keyFilterShowParticipating: Key<Boolean>
+
+  /** The key which indicates if only administrating tournaments should be displayed. */
+  private val keyFilterShowAdministrating: Key<Boolean>
+
+  init {
+    val (prefs, factory) = dataStoreFactory.createPreferencesDataStore()
```
[the remark this answers is not in the document]
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt:57]: (discussed in person)

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/tournaments/TournamentFacade.kt:156]: ah true ^^

File: mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulContestsScreen.kt [hunk 5]
```diff
@@ -84,13 +84,18 @@ fun StatefulTournamentScreen(
     currentUser: AuthenticatedUser,
     onTournamentClick: (TournamentReference) -> Unit,
     onNewContestClickAction: () -> Unit,
+    onFilterClick: () -> Unit,
```
Comment (matt989253) [inline mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulContestsScreen.kt:87]: Missing documentation.

Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulContestsScreen.kt:88]: Fixed in 9a134a69295616820a0cccaeb459af8d3fd831ed.

File: mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulHome.kt [hunk 6]
```diff
@@ -125,10 +130,20 @@ fun StatefulHome(
             currentUser = user,
             onTournamentClick = openTournament,
             onNewContestClickAction = { controller.navigate(CreateTournamentRoute) },
+            onFilterClick = { controller.navigate(TournamentFiltersRoute) },
```
[the remark this answers is not in the document]
Comment (matt989253) [reply mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulHome.kt:133]: My bad, I got baited by Github's line hiding.

**Review by barmettlerl — APPROVED** [review]
LGTM

**Pull request merged**